test/e2e/cni/helpers.go

```diff
@@ -0,0 +1,445 @@
+func GetTesterPodNodeName(f *framework.Framework, nsName string, podName string) (string, error) {
+	testerPod, err := f.ClientSet.CoreV1().Pods(nsName).Get(podName, metav1.GetOptions{})
+	return testerPod.Spec.NodeName, err
+}
+
+// GetTestNodes gets the nodes that are not running the tests
+// TODO handle node status
+func GetTestNodes(f *framework.Framework, testerNodeName string) ([]corev1.Node, error) {
+	var testNodes []corev1.Node
+
+	nodesList, err := f.ClientSet.CoreV1().Nodes().List(metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	if len(nodesList.Items) == 0 {
+		return nil, errors.New("No nodes found")
+	}
+
+	for _, node := range nodesList.Items {
+		if testerNodeName != node.Name {
+			log.Debugf("Found test node (%s)", node.Name)
+			testNodes = append(testNodes, node)
+		}
+	}
+	return testNodes, nil
+}
+
+// NodeCoreDNSCount returns the number of KubeDNS or CoreDNS pods running on the node
+func NodeCoreDNSCount(f *framework.Framework, nodeName string) (int, error) {
+	var err error
+	var count int
+	var podList *corev1.PodList
+	// Find nodes running coredns
+	listOptions := metav1.ListOptions{
+		LabelSelector: "k8s-app=kube-dns",
+	}
+	podList, err = f.ClientSet.CoreV1().Pods("kube-system").List(listOptions)
+	if err != nil {
+		return 0, err
+	}
+	for _, pod := range podList.Items {
+		if nodeName == pod.Spec.NodeName {
+			count++
+		}
+	}
+	return count, nil
+}
+
+// GetInstanceLimits gets instance ENI and IP limits
+func GetInstanceLimits(f *framework.Framework, nodeName string) (int, int, error) {
+	filterName := "private-dns-name"
+	describeInstancesInput := &ec2.DescribeInstancesInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   &filterName,
+				Values: []*string{&nodeName},
+			},
+		},
+	}
+	instance, err := f.Cloud.EC2().DescribeInstances(describeInstancesInput)
+	if err != nil {
+		return 0, 0, err
+	}
+	if len(instance.Reservations) < 1 {
+		return 0, 0, errors.New("No instance reservations found")
+	}
+	if len(instance.Reservations[0].Instances) < 1 {
+		return 0, 0, errors.New("No instances found")
+	}
+	instanceType := *(instance.Reservations[0].Instances[0].InstanceType)
+
+	return awsutils.InstanceENIsAvailable[instanceType],
+		awsutils.InstanceIPsAvailable[instanceType] - 1, nil
+}
+
+// GetNodeInternalIP gets a node's internal IP address
+func GetNodeInternalIP(node corev1.Node) (string, error) {
+	if len(node.Status.Addresses) == 0 {
+		return "", fmt.Errorf("No addresses found for node (%s)", node.Name)
+	}
+
+	var internalIP string
+	for _, address := range node.Status.Addresses {
+		if address.Type == corev1.NodeInternalIP {
+			internalIP = address.Address
+		}
+	}
+	return internalIP, nil
+}
+
+// WaitForASGDesiredCapacity ensures the autoscaling groups have the requested desired capacity and the nodes become ready
+func WaitForASGDesiredCapacity(ctx context.Context, f *framework.Framework, nodes []corev1.Node, desiredCapacity int) error {
```

Comment:
I'm surprised that the AWS EC2 go SDK doesn't have a wait function for this already.

Reply:
It's broken. https://github.com/aws/aws-sdk-go/issues/2478

Reply:
Also this waits for nodes to be ready too.